In [106]:
import numpy as np 
import matplotlib.pyplot as plt 
import pandas as pd 

In [107]:
import pandas as pd

# Setze die globale Einstellung für die Anzeige von numerischen Werten
pd.set_option('display.float_format', '{:.2f}'.format)

# Pfad zur Excel-Datei
path_to_excel = "../data/EVS4_20140118_dataV9.xlsx"

# Lade die Excel-Datei
df = pd.read_excel(path_to_excel, sheet_name="DatasetV9")
df

BNR_NNR     cTOTEXn     cTOTEXs  cTOTEXs_RP3  cTOTEXn_RP3       BNR  \
0    10000110_1 12518828.21 12498882.13  11228047.16  10982980.74  10000110   
1    10000127_1 13683481.37 13693533.90  13163748.75  12997565.00  10000127   
2    10000140_1 12221171.36 12401667.63  10428259.02  10349263.70  10000140   
3    10000148_1  7565416.22  7593657.07   7223779.68   7208621.86  10000148   
4    10000150_1  9376233.04  9742012.61   9314366.51   8542815.76  10000150   
..          ...         ...         ...          ...          ...       ...   
189  10011648_1 15971572.23 16077552.08          NaN          NaN  10011648   
190  10012047_1 88524836.30 86727436.78          NaN          NaN  10012047   
191  10012048_1 11155293.27 11142850.36          NaN          NaN  10012048   
192  10012071_1 89000544.97 83890615.90          NaN          NaN  10012071   
193  10012361_1 15772133.91 15413038.23          NaN          NaN  10012361   

     dmu   dDateData                                     dmuName  \
0      1  21.10.2022                 Stadtwerke Schweinfurt GmbH   
1      2  13.10.2023                                    EWR GmbH   
2      3  09.10.2023  Vereinigte Wertach-Elektrizitätswerke GmbH   
3      4  01.12.2022                         Stadtwerke Elmshorn   
4      5  05.08.2022                  Überlandwerk Leinetal GmbH   
..   ...         ...                                         ...   
189  193  06.06.2023                               CPM Netz GmbH   
190  195  07.10.2022                        Stuttgart Netze GmbH   
191  196  10.10.2023                              LeineNetz GmbH   
192  197  10.10.2023                         e-netz Südhessen AG   
193  198  13.10.2023                     SWTE Netz GmbH & Co. KG   

     yConnections.N1  yConnections.N2  yConnections.N3  yConnections.N4  \
0                  0                0                3               29   
1                  0                0                0                0   
2                  0                0                0               14   
3                  0                0                0                0   
4                  0                0                0               42   
..               ...              ...              ...              ...   
189                0                0               23              236   
190                0                0               30              766   
191                0                0                0                0   
192                0                0               33              394   
193                0                0                0                0   

     yConnections.N5  yConnections.N6  yConnections.N7  yConnections.tot  \
0                481             3489            14321             18323   
1                792             7186            23035             31013   
2                488             4982            20258             25742   
3                314             1970            17158             19442   
4                469             3451            21861             25823   
..               ...              ...              ...               ...   
189              307             1060              505              2131   
190             2406            13554            82752             99508   
191              700             3132            30653             34485   
192             3782            24962           179633            208804   
193             1178             4638            39399             45215   

     yConnections.streetlights.dir.N7  yConnections.streetlights.N7  \
0                                  11                           112   
1                                   0                           536   
2                                   0                           243   
3                                   0                            37   
4                                   0                           175   
..

# NA-Values

In [108]:
missing_rows = df.loc[df.isnull().sum(axis=1) > 9]
missing_rows

Empty DataFrame
Columns: [BNR_NNR, cTOTEXn, cTOTEXs, cTOTEXs_RP3, cTOTEXn_RP3, BNR, dmu, dDateData, dmuName, yConnections.N1, yConnections.N2, yConnections.N3, yConnections.N4, yConnections.N5, yConnections.N6, yConnections.N7, yConnections.tot, yConnections.streetlights.dir.N7, yConnections.streetlights.N7, yConnections.streetlights.N6, yConnections.streetlights.sum, yConnections.excl.streetlights.tot, yConnections.cus.N1, yConnections.cus.N2, yConnections.cus.N3, yConnections.cus.N4, yConnections.cus.N5, yConnections.cus.N6, yConnections.cus.N7, yConnections.cus.tot, yConnections.other.dso.lower.N1to6.sum, yConnections.other.dso.same.tot, yInjectionPoints.N1, yInjectionPoints.N2, yInjectionPoints.N3, yInjectionPoints.N4, yInjectionPoints.N5, yInjectionPoints.N6, yInjectionPoints.N7, yInjectionPoints.renewables.N1, yInjectionPoints.renewables.N2, yInjectionPoints.renewables.N3, yInjectionPoints.renewables.N4, yInjectionPoints.renewables.N5, yInjectionPoints.renewables.N6, yInjectionPoints.renewables.N7, yInjectionPoints.renewables.solar.N1, yInjectionPoints.renewables.solar.N2, yInjectionPoints.renewables.solar.N3, yInjectionPoints.renewables.solar.N4, yInjectionPoints.renewables.solar.N5, yInjectionPoints.renewables.solar.N6, yInjectionPoints.renewables.solar.N7, yInjectionPoints.renewables.wind.N1, yInjectionPoints.renewables.wind.N2, yInjectionPoints.renewables.wind.N3, yInjectionPoints.renewables.wind.N4, yInjectionPoints.renewables.wind.N5, yInjectionPoints.renewables.wind.N6, yInjectionPoints.renewables.wind.N7, yInjectionPoints.renewables.biomass.N1, yInjectionPoints.renewables.biomass.N2, yInjectionPoints.renewables.biomass.N3, yInjectionPoints.renewables.biomass.N4, yInjectionPoints.renewables.biomass.N5, yInjectionPoints.renewables.biomass.N6, yInjectionPoints.renewables.biomass.N7, yInjectionPoints.renewables.hydro.N1, yInjectionPoints.renewables.hydro.N2, yInjectionPoints.renewables.hydro.N3, yInjectionPoints.renewables.hydro.N4, yInjectionPoints.renewables.hydro.N5, yInjectionPoints.renewables.hydro.N6, yInjectionPoints.renewables.hydro.N7, yInjectionPoints.KWKG.N1, yInjectionPoints.KWKG.N2, yInjectionPoints.KWKG.N3, yInjectionPoints.KWKG.N4, yInjectionPoints.KWKG.N5, yInjectionPoints.KWKG.N6, yInjectionPoints.KWKG.N7, yInjectionPoints.other.N1, yInjectionPoints.other.N2, yInjectionPoints.other.N3, yInjectionPoints.other.N4, yInjectionPoints.other.N5, yInjectionPoints.other.N6, yInjectionPoints.other.N7, yInjectionPoints.tot, yInjectionPoints.N5to7.sum, yInjectionPoints.dec.sum, yInjectionPoints.renewables.tot, yInjectionPoints.renewables.solar.tot, yInjectionPoints.renewables.wind.tot, yConnections.incl.inj.N1, yConnections.incl.inj.N2, yConnections.incl.inj.N3, yConnections.incl.inj.N4, yConnections.incl.inj.N5, yConnections.incl.inj.N6, ...]
Index: []

[0 rows x 904 columns]

In a single row there are at most 9 columns missing, so there is no need to drop any observation.

In [109]:
missing_values = df.isnull().sum()
missing_values_sorted = missing_values.sort_values(ascending=False)
missing_values_sorted.head(20)

yRelativeLowerPower.scaled.corr.N4                    67
yRelativeLowerPower.N4                                67
yRelativeLowerPower.scaled.N4                         67
rUsageHours.delivered.N4                              20
rUsageHours.injection.N4                              20
rMeters.read.per.connection.cus.N3                    14
rMeters.read.per.connection.cus.N4                    11
cTOTEXn_RP3                                            9
cTOTEXs_RP3                                            9
rUsageHours.delivered.N2                               7
rUsageHours.injection.N2                               7
rInstalledPower.per.peakinj.N4                         4
rMeters.read.per.connection.cus.N6                     4
rInstalledPower.renewables.per.point.renewables.N4     3
rInstalledPower.renewables.solar.per.point.N3          3
rInstalledPower.renewables.solar.per.point.N4          3
rInstalledPower.per.peakinj.N2                         3
rInstalledPower.renewables.wind

These are the 20/904 variables with the most missing values. Considering the low number of observations (194) we will drop yRelativeLowerPower.scaled.corr.N4, yRelativeLowerPower.N4 and yRelativeLowerPower.scaled.N4 which have more than 30% of their values missing.  
We also drop irrelevant columns for the prediction such as name, date of data collection and the TOTEX-variables besides cTOTEXn.

In [110]:
pd.set_option("display.max_rows", 100)
(df == 0).sum().sort_values(ascending=False).head(50)

yInstalledPower.renewables.solar.reducedAPFI.N1       194
yMeters.cp.ctrl.N2                                    194
yInstalledPower.renewables.hydro.N2                   194
yInstalledPower.renewables.hydro.N1                   194
yInstalledPower.renewables.biomass.N2                 194
yInstalledPower.renewables.biomass.N1                 194
yInstalledPower.renewables.wind.N1                    194
rInstalledPower.renewables.solar.per.point.N2         194
yInstalledPower.renewables.solar.N2                   194
rInstalledPower.renewables.solar.per.point.N1         194
yMeters.flatrate.N1                                   194
yMeters.flatrate.N2                                   194
yMeters.flatrate.N3                                   194
yInstalledPower.renewables.solar.N1                   194
yInstalledPower.renewables.N1                         194
ySwitchingstations.N1                                 194
yTransformers.power.ront.reserve.own.N7               194
ySubstations.N

There are many extremely sparse variables, with mostly 0 values. We set a rather lax condition for removing these (>90% missing values), as we will be using models that work well with sparse features such as Random Forest or Lasso Regression and otherwise use feature selection methods.

In [111]:
# Drop irrelevant columns for prediction and with more than 30% missing values
columns_to_drop = ['cTOTEXs', 'cTOTEXs_RP3', 'cTOTEXn_RP3', 'NameOrg', 'NameShort', 'dmuName', 'dmu', 'dDateData', 'BNR', 'BNR_NNR', 'yRelativeLowerPower.scaled.corr.N4', 'yRelativeLowerPower.N4', 'yRelativeLowerPower.scaled.N4']
threshold = 0.9 * len(df)

for col in df.columns:
    if (df[col] == 0).sum() > threshold:
        columns_to_drop.append(col)

df_cleaned = df.drop(columns=columns_to_drop)
print("number of columns to drop: {}".format(len(columns_to_drop)))


number of columns to drop: 327


There are no categorical variables to one-hot-encode.

In [112]:
from sklearn.model_selection import train_test_split
import pandas as pd

# Split the data into training and validation sets (90% training, 10% validation)
df_train, df_val = train_test_split(df_cleaned, test_size=0.1, random_state=42)

# Check the size of the split datasets
print("Size of df_train:", df_train.shape)
print("Size of df_val:", df_val.shape)

Size of df_train: (174, 577)
Size of df_val: (20, 577)


## KNN for Imputation of remaining NA values

In [113]:
print(df_train.isnull().sum().sum())
print(df_val.isnull().sum().sum())

92
6


In [114]:
from sklearn.model_selection import train_test_split
from sklearn.impute import KNNImputer
from sklearn.preprocessing import StandardScaler
import pandas as pd

# Initialize the StandardScaler
scaler = StandardScaler()

# Fit the scaler on the training data and transform both training and validation data
df_train_scaled = scaler.fit_transform(df_train)
df_val_scaled = scaler.transform(df_val)

# Initialize the KNN Imputer
imputer = KNNImputer(n_neighbors=3)

# Fit the imputer on the training data and transform both training and validation data
df_train_imputed_scaled = imputer.fit_transform(df_train_scaled)
df_val_imputed_scaled = imputer.transform(df_val_scaled)

# Inverse transform to convert the data back to the original scale
df_train_imputed = scaler.inverse_transform(df_train_imputed_scaled)
df_val_imputed = scaler.inverse_transform(df_val_imputed_scaled)

# Convert the results back to DataFrames
df_train_imputed = pd.DataFrame(df_train_imputed, columns=df_train.columns)
df_val_imputed = pd.DataFrame(df_val_imputed, columns=df_val.columns)


In [115]:
from sklearn.model_selection import train_test_split
from sklearn.impute import KNNImputer
from sklearn.preprocessing import StandardScaler
import pandas as pd

# Initialize the StandardScaler
scaler = StandardScaler()

# Fit the scaler on the training data and transform both training and validation data
df_train_scaled = scaler.fit_transform(df_train)
df_val_scaled = scaler.transform(df_val)

# Initialize the KNN Imputer
imputer = KNNImputer(n_neighbors=3)

# Fit the imputer on the training data and transform both training and validation data
df_train_imputed_scaled = imputer.fit_transform(df_train_scaled)
df_val_imputed_scaled = imputer.transform(df_val_scaled)

# Inverse transform to convert the data back to the original scale
df_train_imputed = scaler.inverse_transform(df_train_imputed_scaled)
df_val_imputed = scaler.inverse_transform(df_val_imputed_scaled)

# Convert the results back to DataFrames
df_train_imputed = pd.DataFrame(df_train_imputed, columns=df_train.columns)
df_val_imputed = pd.DataFrame(df_val_imputed, columns=df_val.columns)


Number of missing values in the training set after imputation: 0
Number of missing values in the validation set after imputation: 0


# Outlier detection

There are many variables that have both many 0 values and many very large values due to very different network structures for each of the providers. Therefore, a simple z-score approach on the full dataset faces issues with identifying outliers as the 0s skew the distribution heavily. Instead, we look only at data points larger than the mean and use the z-score threshold of 3 on this subset to identify outliers.

In [116]:
import pandas as pd
import numpy as np
from scipy import stats

pd.set_option("display.max_rows", 50)

# Define function to detect outliers in the value range larger than the mean of each column
def detect_high_value_outliers(df, z_thresh=3):
    outlier_indices = set()
    outlier_columns = []

    for col in df.columns:
        thresh_value = df[col].mean()

        # Filter above mean
        high_values = df[df[col] > thresh_value][col]
        
        # Calculate z-scores for this subset
        high_value_z_scores = np.abs(stats.zscore(high_values.dropna()))
        
        # Identify outliers based on z-threshold
        high_value_outliers = high_values[high_value_z_scores > z_thresh]
        
        if not high_value_outliers.empty:
            outlier_indices.update(high_value_outliers.index.tolist())
            outlier_columns.append(col)
    
    # Get unique outlier indices
    outlier_indices = list(outlier_indices)
    
    # DataFrame with only the outliers and the columns where outliers were found
    outlier_df = df.loc[outlier_indices, outlier_columns]
    
    return outlier_df


In [117]:
# Apply outlier detection to the training set
train_outliers = detect_high_value_outliers(df_train_imputed, z_thresh=3)
print("Training set outliers:")
train_outliers

Training set outliers:


cTOTEXn  yConnections.N3  yConnections.N4  yConnections.N5  \
2   230256581.35           137.00           753.00         10674.00   
132  98561074.71            31.00           463.00          2674.00   
134 105242001.31            27.00           807.00          3471.00   
136 110704579.91            68.00           380.00          4320.00   
8    23457592.09             0.00            69.00          1388.00   
..           ...              ...              ...              ...   
117  83505453.44            35.00           642.00          3901.00   
118  16849182.65             0.00             3.00           571.00   
120   6939548.78             0.00             0.00           279.00   
122 374258826.89            91.00          3197.00         11462.00   
123   8689687.59             0.00             0.00           438.00   

     yConnections.N6  yConnections.N7  yConnections.tot  \
2           77004.00        341932.00         430574.00   
132         14365.00         97391.00         114941.00   
134         12683.00         82891.00          99879.00   
136         25976.00        222444.00         253188.00   
8            5148.00         57692.00          64297.00   
..               ...              ...               ...   
117         20500.00         96857.00         121971.00   
118          5049.00         23925.00          29548.00   
120          1993.00         14124.00          16396.00   
122         65964.00        339263.00         419977.00   
123          3359.00         22923.00          26720.00   

     yConnections.streetlights.dir.N7  yConnections.streetlights.N7  \
2                                0.00                       2797.00   
132                          32497.00                       1188.00   
134                           2676.00                       2122.00   
136                              0.00                        449.00   
8                               39.00                        216.00   
..                                ...                           ...   
117                              0.00                       1179.00   
118                              0.00                        463.00   
120                            422.00                        143.00   
122                         189552.00                          0.00   
123                              0.00                        291.00   

     yConnections.streetlights.N6  yConnections.streetlights.sum  \
2                            0.00                        2797.00   
132                        181.00                        1369.00   
134                          0.00                        2122.00   
136                       6833.00                        7282.00   
8                          310.00                         526.00   
..                            ...                            ...   
117                          0.00                        1179.00   
118                        159.00                         622.00   
120                         74.00                         217.00   
122                          0.00                           0.00   
123                          0.00                         291.00   

     yConnections.excl.streetlights.tot  yConnections.cus.N4  \
2                             427777.00                25.00   
132                           113572.00                 9.00   
134                            97757.00                24.00   
136                           245906.00                24.00   
8                              63771.00                 2.00   
..                                  ...                  ...   
117                           120792.00                 2.00   
118                            28926.00                 0.00   
120                            16179.00                 0.00   
122                           419977.00                 6.00   
123                            26429.00                 0.00   

     yCo

In [118]:
# Apply outlier detection to the validation set
val_outliers = detect_high_value_outliers(df_val_imputed, z_thresh=3)
print("Validation set outliers:")
val_outliers

Validation set outliers:


rPeakload.injection.per.Peakload.N4  rMeters.active.share.N5  \
17                                 1.01                     1.00   
10                                 1.63                     1.00   
11                                 1.00                     1.00   

    rUsageHours.injection.N4  rUsageHours.delivered.N4  
17                   7426.96                   7464.84  
10                   3980.28                   6489.14  
11                  15372.65                  15406.80

Even with this approach, in 392 out of 577 columns outliers exist. 

In [119]:
pd.set_option("display.max_columns", 400)

# Display the descriptive statistics for the training set
train_description = df_train_imputed.describe(percentiles=[0.9, 0.95, 0.99])
print("Training set description:")
train_description

Training set description:


cTOTEXn  yConnections.N3  yConnections.N4  yConnections.N5  \
count        174.00           174.00           174.00           174.00   
mean    62022906.12            29.59           300.33          2695.59   
std    154397591.84           109.13           768.20          6887.50   
min      1432414.03             0.00             0.00             0.00   
50%     15578563.37             0.00            63.00           732.50   
90%    103237723.33            63.40           699.50          4418.30   
95%    252181074.22            98.35          1598.15         10866.85   
99%    888241492.70           448.28          3698.65         38489.75   
max   1280552772.14          1106.00          6162.00         56803.00   

       yConnections.N6  yConnections.N7  yConnections.tot  \
count           174.00           174.00            174.00   
mean          13282.56         92076.28         108392.56   
std           32057.43        225482.18         263612.09   
min               0.00             0.00             20.00   
50%            4095.00         24901.50          30700.50   
90%           24298.10        161424.10         196461.50   
95%           61657.45        373504.20         445276.80   
99%          149341.30       1212334.49        1421181.35   
max          292915.00       1953427.00        2241803.00   

       yConnections.streetlights.dir.N7  yConnections.streetlights.N7  \
count                            174.00                        174.00   
mean                            6204.67                        835.85   
std                            32498.72                       2858.12   
min                                0.00                          0.00   
50%                                0.00                        209.00   
90%                             5664.10                       1428.20   
95%                            24543.15                       2215.35   
99%                           190251.84                      13661.54   
max                           325003.00                      28308.00   

       yConnections.streetlights.N6  yConnections.streetlights.sum  \
count                        174.00                         174.00   
mean                         232.39                        1068.24   
std                          977.66                        3341.17   
min                            0.00                           0.00   
50%                            0.00                         296.00   
90%                          345.90                        1778.40   
95%                          705.95                        3927.60   
99%                         4330.56                       16098.02   
max                         9841.00                       29125.00   

       yConnections.excl.streetlights.tot  yConnections.cus.N3  \
count                              174.00               174.00   
mean                            107324.32                 6.56   
std                             261430.74                27.23   
min                                 20.00                 0.00   
50%                              30432.50                 0.00   
90%                             194342.90                 9.40   
95%                             433550.95                20.80   
99%                            1404862.74               170.94   
max                            2231263.00               188.00   

       yConnections.cus.N4  yConnections.cus.N5  yConnections.cus.N6  \
count               174.00               174.00               174.00   
mean                 10.25               713.17               360.63   
std                  59.96              1826.06              1285.45   
min                   0.00                 0.00                 0.00   
50%                   0.00               206.00                56.00   
90%                  15.70              1304.20               621.40   
95%                  22.35              2837.75              16

In [120]:
# Display the descriptive statistics for the validation set
val_description = df_val_imputed.describe(percentiles=[0.9, 0.95, 0.99])
print("Validation set description:")
val_description

Validation set description:


cTOTEXn  yConnections.N3  yConnections.N4  yConnections.N5  \
count        20.00            20.00            20.00            20.00   
mean   78266658.73            59.75           419.75          3461.65   
std   123179932.78           195.22           697.37          5656.44   
min     4936708.06             0.00             0.00           273.00   
50%    25149312.10             0.00            95.50           989.50   
90%   269317157.79           103.50          1748.20          8559.10   
95%   280069566.47           163.45          1980.25         18181.30   
99%   440456773.81           732.69          2196.85         19553.86   
max   480553575.65           875.00          2251.00         19897.00   

       yConnections.N6  yConnections.N7  yConnections.tot  \
count            20.00            20.00             20.00   
mean          16973.45        140438.45         161360.10   
std           24632.80        227374.94         257347.72   
min            1179.00         12344.00          13796.00   
50%            6914.50         41358.00          49353.00   
90%           51715.90        359606.40         423976.30   
95%           73264.85        655168.65         764652.10   
99%           86045.77        821086.53         918219.22   
max           89241.00        862566.00         956611.00   

       yConnections.streetlights.dir.N7  yConnections.streetlights.N7  \
count                             20.00                         20.00   
mean                            2940.45                        842.75   
std                            11035.15                        913.50   
min                                0.00                          0.00   
50%                                0.00                        367.00   
90%                             2398.90                       1905.10   
95%                             9775.80                       2806.35   
99%                            41343.16                       2811.67   
max                            49235.00                       2813.00   

       yConnections.streetlights.N6  yConnections.streetlights.sum  \
count                         20.00                          20.00   
mean                          36.85                         879.60   
std                          112.83                         892.59   
min                            0.00                          35.00   
50%                            0.00                         475.00   
90%                           43.10                        1905.80   
95%                          252.80                        2813.70   
99%                          416.96                        2824.34   
max                          458.00                        2827.00   

       yConnections.excl.streetlights.tot  yConnections.cus.N3  \
count                               20.00                20.00   
mean                            160480.50                14.15   
std                             257104.98                42.42   
min                              13681.00                 0.00   
50%                              48836.00                 0.00   
90%                             422348.30                29.40   
95%                             764535.50                49.25   
99%                             916862.30               159.45   
max                             954944.00               187.00   

       yConnections.cus.N4  yConnections.cus.N5  yConnections.cus.N6  \
count                20.00                20.00                20.00   
mean                 29.25               725.95               231.30   
std                  85.67              1043.02               380.79   
min                   0.00                68.00                 0.00   
50%                   1.00               261.00                81.50   
90%                  45.80              1881.90               524.50   
95%                 120.55              2983.40               938.30   


When looking closer at the different quantiles in the dataset, we can see that there are some network providers with much higher capacity and cost. These are real values, but they are still sometimes magnitudes larger than other networks. Completely removing them or capping/clipping them could result in loss of information and relationship. Some models like Random Forest and Gradient Boosting are less sensitive to outliers, but especially regression methods are largely influenced by outliers. Therefore, we propose to limit the effect by trying out different preprocessing for the models affected: log-transforming (some) variables or clipping the values to the 0.99-quantile. We will also run separate regression methods for larger or smaller network providers. We will also investigate, whether clusters of similar network providers exist and whether it makes sense to predict different models for different clusters, assuming they each have different cost drivers in their network.

In [121]:
# Save the training set to a CSV file
train_set_path = "../data/trainingset.csv"
df_train_imputed.to_csv(train_set_path, index=False)

print(f"Training set saved to {train_set_path}")

# Save the validation set to a CSV file
val_set_path = "../data/validationset.csv"
df_val_imputed.to_csv(val_set_path, index=False)

print(f"Validation set saved to {val_set_path}")

Training set saved to ../data/trainingset.csv
Validation set saved to ../data/validationset.csv


In [123]:
import numpy as np
from scipy.stats import skew
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# Function to detect and apply log or polynomial transformation if necessary
def apply_transformation(df_train, df_val, feature, target, degree=2, skewness_threshold=0.5, improvement_threshold=0.01):
    skewness = skew(df_train[feature].dropna())
    
    if skewness > skewness_threshold:
        # Apply log transformation
        df_train[feature] = np.log1p(df_train[feature])
        df_val[feature] = np.log1p(df_val[feature])
        new_col_name = feature + "_log"
        df_train.rename(columns={feature: new_col_name}, inplace=True)
        df_val.rename(columns={feature: new_col_name}, inplace=True)
        return new_col_name, "log"
    
    # Polynomial regression
    X_train = df_train[[feature]]
    y_train = df_train[target]
    
    # Linear regression
    lin_reg = LinearRegression()
    lin_reg.fit(X_train, y_train)
    y_pred_lin = lin_reg.predict(X_train)
    
    # Polynomial regression
    poly = PolynomialFeatures(degree=degree)
    X_poly_train = poly.fit_transform(X_train)
    poly_reg = LinearRegression()
    poly_reg.fit(X_poly_train, y_train)
    y_pred_poly = poly_reg.predict(X_poly_train)
    
    # R-squared comparison
    r2_lin = r2_score(y_train, y_pred_lin)
    r2_poly = r2_score(y_train, y_pred_poly)
    
    # Apply polynomial transformation if it improves the R-squared significantly
    if r2_poly - r2_lin > improvement_threshold:
        df_train[feature] = y_pred_poly
        X_val = df_val[[feature]]
        X_poly_val = poly.transform(X_val)
        df_val[feature] = poly_reg.predict(X_poly_val)
        new_col_name = feature + "_poly"
        df_train.rename(columns={feature: new_col_name}, inplace=True)
        df_val.rename(columns={feature: new_col_name}, inplace=True)
        return new_col_name, "poly"
    
    return feature, None

# Apply transformations to training and validation sets
log_transformed_columns = 0
polynomial_transformed_columns = 0
transformed_features = {}

features = df_train_imputed.columns.drop(target_column)

for feature in features:
    new_name, transformation_type = apply_transformation(df_train_imputed, df_val_imputed, feature, target_column)
    if transformation_type == "log":
        log_transformed_columns += 1
    elif transformation_type == "poly":
        polynomial_transformed_columns += 1
    transformed_features[feature] = new_name

print(f"Number of columns log-transformed: {log_transformed_columns}")
print(f"Number of columns polynomial-transformed: {polynomial_transformed_columns}")


Number of columns log-transformed: 550
Number of columns polynomial-transformed: 6
Transformations applied to columns:
       cTOTEXn  yConnections.N3_log  yConnections.N4_log  yConnections.N5_log  \
0  65381795.52                 3.14                 5.62                 7.75   
1  87582629.71                 3.33                 6.47                 8.30   
2 230256581.35                 4.93                 6.63                 9.28   
3  12975571.18                 0.00                 4.72                 6.47   
4  36097410.97                 2.56                 5.38                 7.04   

   yConnections.N6_log  yConnections.N7_log  yConnections.tot_log  \
0                 9.10                11.35                 11.48   
1                 9.68                11.38                 11.59   
2                11.25                12.74                 12.97   
3                 7.99                 9.87                 10.05   
4                 9.14                10.75      

In [127]:
# Save the training set to a CSV file
train_set_path = "../data/trainingsettransformed.csv"
df_train_imputed.to_csv(train_set_path, index=False)

print(f"Training set saved to {train_set_path}")

# Save the validation set to a CSV file
val_set_path = "../data/validationsettransformed.csv"
df_val_imputed.to_csv(val_set_path, index=False)

print(f"Validation set saved to {val_set_path}")

Training set saved to ../data/trainingsettransformed.csv
Validation set saved to ../data/validationsettransformed.csv
